test/test_katello/test_deb_tracer.py

```diff
@@ -0,0 +1,144 @@
+NEEDRESTART-SVC: rsyslog.service
+NEEDRESTART-SVC: ssh.service
+NEEDRESTART-SVC: systemd-journald.service
+NEEDRESTART-SVC: systemd-logind.service
+NEEDRESTART-SVC: systemd-timesyncd.service
+NEEDRESTART-SVC: unattended-upgrades.service
+NEEDRESTART-SVC: user@0.service """
+
+        output_list = needrestart_output.split("\n")
+        m.return_value = needrestart_output
+        self.assertEqual(needrestart(), output_list)
+        self.assertEqual(m.call_args.args, (["needrestart", "-b"],))
+        self.assertEqual(m.call_args.kwargs, {"universal_newlines": True})
+        self.assertEqual(len(collect_apps()), 12)
+
+    @patch.object(subprocess, "check_output")
+    def test_needrestart_exception_file_not_found(self, m):
+        m.side_effect = OSError()
+        with self.assertRaises(SystemExit) as cm:
+            needrestart()
+        the_exception = cm.exception
+        self.assertEqual(the_exception.code, 1)
+
+
+class TestCollectApps(unittest.TestCase):
+    @unittest.skipIf(sys.version_info[0] == 3, "not compatible with Python 3")
+    @patch("katello.deb_tracer.spawn.find_executable")
+    def test_has_needrestart_not_python2(self, m):
+        m.return_value = None
+        self.assertEqual(has_needrestart(), False)
+        self.assertEqual(m.call_args.args, ("needrestart",))
+
+    @unittest.skipIf(sys.version_info[0] < 3, "not compatible with Python 2")
+    @patch("katello.deb_tracer.which")
+    def test_has_needrestart_not_python3(self, m):
+        m.return_value = None
+        self.assertEqual(has_needrestart(), False)
+        self.assertEqual(m.call_args.args, ("needrestart",))
+
+    @unittest.skipIf(sys.version_info[0] == 3, "not compatible with Python 3")
+    @patch("katello.deb_tracer.spawn.find_executable")
+    def test_has_needrestart_ok_python2(self, m):
+        m.return_value = "/usr/sbin/needrestart"
+        self.assertEqual(has_needrestart(), True)
+        self.assertEqual(m.call_args.args, ("needrestart",))
+
+    @unittest.skipIf(sys.version_info[0] < 3, "not compatible with Python 2")
+    @patch("katello.deb_tracer.which")
+    def test_has_needrestart_ok_python3(self, m):
+        m.return_value = "/usr/sbin/needrestart"
+        self.assertEqual(has_needrestart(), True)
+        self.assertEqual(m.call_args.args, ("needrestart",))
+
+
+class TestUseNeedRestart(unittest.TestCase):
+    @patch("katello.deb_tracer.needrestart")
+    def test_system_reboot(self, m):
+        output = [
+            "NEEDRESTART-VER: 3.5",
+            "NEEDRESTART-KCUR: 5.10.0-7-amd64",
+            "NEEDRESTART-KEXP: 5.10.0-10-amd64",
+            "NEEDRESTART-KSTA: 3"
+        ]
+        m.return_value = output
+        apps = use_needrestart()
+        self.assertEqual(type(apps), list)
+        self.assertTrue(len(apps) > 0)
+        self.assertIsInstance(apps[0], AptTracerApp)
+        self.assertTrue(apps[0].name == "kernel" and apps[0].helper == "Please restart your system" and apps[0].type == "static")
```

Comment:
I would only check for name and  type. In all cases.

Reply:
In addition to that I would make it multiple assertions. This big boolean is hard to debug when it fails. I'd make it `assertEqual`.